### PR TITLE
🎉 Source Shopify - change the default stream cursor_field to 'updated_at' where possible

### PR DESCRIPTION
**Pull request by bazarnov**
## What
#4346 - Source-Shopify migrate from since_id to datetime

## How
- changed the `request_parameters` in `ShopifyStream class`
- changed the `request_parameters` in the `IncrementalShopifyStream class`
- modified code for all stream classes
- added state comparison to next classes: `OrderRefunds, OrderRisks, Transactions, DiscountCodes` to catch the latest records

## Pre-merge Checklist
Expand the checklist which is relevant for this PR. 

<details><summary> <strong> Connector checklist </strong> </summary>
<p>

- [x] Issue acceptance criteria met
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [X] Unit & integration tests added as appropriate (and are passing)
    `./gradlew :airbyte-integrations:connectors:source-shopify:integrationTest`.
- [ ] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
- [ ] Code reviews completed
- [X] Documentation updated 
    - [X] Updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [X] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [X] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [ ] Build is successful
- [ ] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 1]
```diff
@@ -170,74 +165,140 @@ class Metafields(IncrementalShopifyStream):
...
-            yield from super().read_records(stream_slice={"order_id": data["id"]}, **kwargs)
+            refund_records = super().read_records(stream_slice={"order_id": data["id"]}, **kwargs)
+            # Getting records >= state
+            if stream_state:
+                # For the case where the order has multiple risks
+                for record in refund_records:
+                    if record[self.cursor_field] >= stream_state.get(self.cursor_field):
+                        yield record
+            else:
+                yield from refund_records
```
Comment (yaroslav-dudar) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:273]: I see similar code in multiple places. Could you create a create a function and call it in multiple spots ?

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 2]
```diff
@@ -119,19 +124,14 @@ class Orders(IncrementalShopifyStream):
...
         return f"{self.data_field}.json"
 
-    def request_params(self, stream_state=None, next_page_token: Mapping[str, Any] = None, **kwargs) -> MutableMapping[str, Any]:
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
         stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, next_page_token=next_page_token, **kwargs)
+        params.update({"status": "any"})
         if next_page_token:
```
Comment (yaroslav-dudar) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:133]: I think it's better to
```
if not next_page_token:
    params.update({"status": "any"})
```

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py]: Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 2 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py]: Done.

**Review by yaroslav-dudar — APPROVED** [review]

**Review by TymoshokDmytro — APPROVED** [review]

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 3]
```diff
@@ -170,74 +174,117 @@ class Metafields(IncrementalShopifyStream):
...
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, next_page_token=next_page_token, **kwargs)
+        if not stream_state:
+            params.pop("created_at_min", None)
+            params.update({"published_at_min": self.start_date})
+        if next_page_token:
+            params.pop("published_at_min", None)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:193]: The pagination will not work according to [this documentation](https://shopify.dev/api/usage/pagination-rest) and practice with tests.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 4]
```diff
@@ -119,19 +135,13 @@ class Orders(IncrementalShopifyStream):
...
-                "limit": self.limit,
-                "order": f"{self.primary_key} asc",
-                "created_at_min": self.start_date,
-                "status": "any",
-                # Add state parameter "since_id" for incremental refresh
-                "since_id": stream_state.get(self.cursor_field),
-            }
+        params = super().request_params(stream_state=stream_state, next_page_token=next_page_token, **kwargs)
+        if not next_page_token:
+            params.update({"status": "any"})
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:144]: No.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 5]
```diff
@@ -170,74 +174,117 @@ class Metafields(IncrementalShopifyStream):
...
     def path(self, **kwargs) -> str:
         return f"{self.data_field}.json"
 
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, next_page_token=next_page_token, **kwargs)
+        if not stream_state:
+            params.pop("created_at_min", None)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:190]: This is for full-refresh or the very first incremental refresh, before the state object is created, to capture the data from the time it's created, after that we use the `updated_at` where it possible to capture the updated data.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 6]
```diff
@@ -170,74 +174,117 @@ class Metafields(IncrementalShopifyStream):
...
+            params.update({"published_at_min": self.start_date})
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:191]: is this a bottleneck? how much does it impact the code execution?

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 7]
```diff
@@ -261,12 +308,16 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
...
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs
+    ) -> Iterable[Mapping[str, Any]]:
+        stream_state = stream_state or {}
         price_rules_stream = PriceRules(
             authenticator=self.authenticator, shop=self.shop, start_date=self.start_date, api_password=self.api_password
         )
         for data in price_rules_stream.read_records(sync_mode=SyncMode.full_refresh):
-            yield from super().read_records(stream_slice={"price_rule_id": data["id"]}, **kwargs)
+            discount_slice = super().read_records(stream_slice={"price_rule_id": data["id"]}, **kwargs)
+            yield from super().parse_stream_slice(stream_state=stream_state, records_slice=discount_slice)
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:320]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 8]
```diff
@@ -261,12 +308,16 @@ def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
         price_rule_id = stream_slice["price_rule_id"]
         return f"price_rules/{price_rule_id}/{self.data_field}.json"
 
-    def read_records(self, stream_slice: Optional[Mapping[str, Any]] = None, **kwargs) -> Iterable[Mapping[str, Any]]:
+    def read_records(
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:311]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 9]
```diff
@@ -41,31 +41,31 @@ class ShopifyStream(HttpStream, ABC):
...
             return None
 
     def request_params(
-        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+        self,
+        next_page_token: Mapping[str, Any] = None,
+        **kwargs,
     ) -> MutableMapping[str, Any]:
         if next_page_token:
             params = {"limit": self.limit, **next_page_token}
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:71]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:191]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 10]
```diff
@@ -170,74 +174,117 @@ class Metafields(IncrementalShopifyStream):
...
+    def request_params(
+        self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
         stream_state = stream_state or {}
+        params = super().request_params(stream_state=stream_state, next_page_token=next_page_token, **kwargs)
+        params.pop("order", None)
+        params.pop("created_at_min", None)
+        if stream_state:
+            params.pop("updated_at_min", None)
+            params.update({"since_id": stream_state.get(self.cursor_field)})
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:227]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:190]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 11]
```diff
@@ -86,25 +86,41 @@ def data_field(self) -> str:
...
-    cursor_field = "id"
+    # Setting the default cursor field for all streams
+    cursor_field = "updated_at"
 
     def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
-        return {self.cursor_field: max(latest_record.get(self.cursor_field, 0), current_stream_state.get(self.cursor_field, 0))}
+        return {self.cursor_field: max(latest_record.get(self.cursor_field, ""), current_stream_state.get(self.cursor_field, ""))}
 
-    def request_params(self, stream_state=None, **kwargs):
+    def request_params(self, stream_state: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None, **kwargs):
```
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:102]: Resolved.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:144]: Resolved.

**Review by Saiful79NZ — APPROVED** [review]

**Comment by bazarnov** [comment]
Closing this branch, moved changes to bazarnov/4346-shopify-migrate-to-datimes, because this branch could not be merged with master.

**Pull request closed without merging**